### PR TITLE
[Role] `az ad/role`: Azure AD Graph API to Microsoft Graph API migration

### LINKED ISSUE #12946: Microsoft Graph API Support
[**Azure Active Directory Graph API**](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-graph-api) has been deprecated:

> We strongly recommend that you use [Microsoft Graph](https://developer.microsoft.com/graph) instead of Azure AD Graph API to access Azure Active Directory (Azure AD) resources. Our development efforts are now concentrated on Microsoft Graph and no further enhancements are planned for Azure AD Graph API. There are a very limited number of scenarios for which Azure AD Graph API might still be appropriate; for more information, see the [Microsoft Graph or the Azure AD Graph](https://developer.microsoft.com/office/blogs/microsoft-graph-or-azure-ad-graph/) blog post and [Migrate Azure AD Graph apps to Microsoft Graph](https://docs.microsoft.com/graph/migrate-azure-ad-graph-overview).

We need to switch Azure CLI to use [**Microsoft Graph API**](https://docs.microsoft.com/en-us/graph/api/overview?toc=./ref/toc.json&view=graph-rest-1.0) instead.

### PR DESCRIPTION
**Pull request by jiasli**
## Description

Resolve #12946

According to https://docs.microsoft.com/en-us/graph/migrate-azure-ad-graph-overview

> Azure Active Directory (Azure AD) Graph is deprecated but won't be retired on June 30, 2022 as [previously announced](https://techcommunity.microsoft.com/t5/azure-active-directory-identity/update-your-applications-to-use-microsoft-authentication-library/ba-p/1257363). Listening closely to your feedback about the challenges of migrating such a critical dependency, we're delaying the retirement date through at least the end of this year, 2022. We’ll provide a retirement update mid-calendar year including releasing more tools to help you to migrate your apps.

This PR migrates the underlying API of `az ad` and `az role` from AD Graph API to Microsoft Graph API.

## Testing Guide

```sh
az ad sp create-for-rbac
az ad app create --display-name mytestapp

# Other commands under
az ad ...
az role ...
```

### CODE REVIEW

**Comment by yonzhan** [comment]
Role

File: src/azure-cli/azure/cli/command_modules/role/__init__.py [hunk 1]
```diff
@@ -24,8 +24,7 @@ def load_command_table(self, args):
         if args and args[0] in ('role', 'ad'):
             from knack.log import get_logger
             logger = get_logger(__name__)
-            logger.warning("The underlying Active Directory Graph API will be replaced by Microsoft Graph API in "
-                           "Azure CLI 2.37.0. "
+            logger.warning("This command or command group has been migrated to Microsoft Graph API. "
                            "Please carefully review all breaking changes introduced during this migration: "
                            "https://docs.microsoft.com/cli/azure/microsoft-graph-migration")
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/__init__.py:29]: The warning message has been updated.

**Review by yonzhan — APPROVED** [review]

**Review by evelyn-ys — APPROVED** [review]

**Comment by jiasli** [comment]
History notes should be added after merging this PR to bypass "Check the Format of Pull Request Title and Content".

**History Notes**

[Role] BREAKING CHANGE: `az az/role`: Migrate the underlying API of `az ad` and `az role` from AD Graph API to Microsoft Graph API. For more details, see [Microsoft Graph migration](https://docs.microsoft.com/en-us/cli/azure/microsoft-graph-migration)

**Review by kairu-ms — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/util.py [hunk 2]
```diff
@@ -978,7 +978,9 @@ def send_raw_request(cli_ctx, method, url, headers=None, uri_parameters=None,  #
         reason = r.reason
         if r.text:
             reason += '({})'.format(r.text)
-        raise CLIError(reason)
+        ex = CLIError(reason)
```
Comment (jsntcy) [inline src/azure-cli-core/azure/cli/core/util.py:981]: What's the actual error type? Do we need use `AzCLIError` (maybe more specific error derived from `AzCLIError`)?

File: src/azure-cli-testsdk/azure/cli/testsdk/utilities.py [hunk 3]
```diff
@@ -48,14 +48,8 @@ def force_progress_logging():
     az_logger.handlers[0].level = old_az_level
 
 
-def _py3_byte_to_str(byte_or_str):
-    import logging
```
Comment (jsntcy) [inline src/azure-cli-testsdk/azure/cli/testsdk/utilities.py:52]: We don't need log here now?

File: src/azure-cli-testsdk/azure/cli/testsdk/utilities.py [hunk 4]
```diff
@@ -157,18 +151,48 @@ def process_response(self, response):
...
 
 
-class AADGraphUserReplacer:
+class MSGraphClientPasswordReplacer(RecordingProcessor):
+    """Replace 'secretText' property in 'addPassword' API's response."""
+
+    PWD_REPLACEMENT = 'replaced-microsoft-graph-password'
+
+    def __init__(self):
+        self._activated = False
```
Comment (jsntcy) [inline src/azure-cli-testsdk/azure/cli/testsdk/utilities.py:160]: What's purpose for `_activated`?

File: src/azure-cli-testsdk/azure/cli/testsdk/utilities.py [hunk 5]
```diff
@@ -157,18 +151,48 @@ def process_response(self, response):
...
+            body = json.loads(response['body']['string'])
+            body['secretText'] = self.PWD_REPLACEMENT
+
+            response['body']['string'] = json.dumps(body)
+            self._activated = False
+
+        return response
+
+
+class MSGraphUserReplacer:
```
Comment (jsntcy) [inline src/azure-cli-testsdk/azure/cli/testsdk/utilities.py:185]: Could you please tell why `MSGraphUserReplacer` doesn't need derive from `RecordingProcessor`?

File: src/azure-cli-core/azure/cli/core/util.py (hunk 2 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/util.py:981]: I just kept the original design. We can refine the error type in the future for `send_raw_request`.

File: src/azure-cli-testsdk/azure/cli/testsdk/utilities.py (hunk 3 above)
Comment (jiasli) [reply src/azure-cli-testsdk/azure/cli/testsdk/utilities.py:52]: Why do we need log here? 🤣 This function simply converts `bytes` to `str`. We seldom care about the warning logged by test framework during a test.

File: src/azure-cli-testsdk/azure/cli/testsdk/utilities.py (hunk 4 above)
Comment (jiasli) [reply src/azure-cli-testsdk/azure/cli/testsdk/utilities.py:160]: This way, we can make this replacer's `process_response` only be called for `/addPassword` API, saving some time. I kept the original design of `GraphClientPasswordReplacer`.

File: src/azure-cli-testsdk/azure/cli/testsdk/utilities.py (hunk 5 above)
Comment (jiasli) [reply src/azure-cli-testsdk/azure/cli/testsdk/utilities.py:185]:
```suggestion
class MSGraphUserReplacer(RecordingProcessor):
```

**Review by jsntcy — APPROVED** [review]

**Review by jsntcy — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Review by calvinhzy — APPROVED** [review]

**Comment by jiasli** [comment]
> Awesome job, great news! Can you estimate when it will be released?

Thanks @lkurzyniec. 😊 You may refer to the release schedule here: https://github.com/Azure/azure-cli/milestone/118

**Pull request merged**